File: app/src/main/java/com/example/db/WorkspaceDao.java
package com.example.db;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

import kotlinx.coroutines.flow.Flow;

@Dao
public interface WorkspaceDao {

    @Query("SELECT * FROM workspaces ORDER BY lastOpenedTimestamp DESC")
    Flow<List<WorkspaceEntity>> getAllWorkspacesFlow();

    @Query("SELECT * FROM workspaces ORDER BY lastOpenedTimestamp DESC")
    List<WorkspaceEntity> getAllWorkspacesSync();

    @Query("SELECT * FROM workspaces WHERE path = :path LIMIT 1")
    WorkspaceEntity getWorkspaceByPath(String path);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertWorkspace(WorkspaceEntity workspace);

    @Query("DELETE FROM workspaces WHERE path = :path")
    void deleteWorkspaceByPath(String path);

    @Query("DELETE FROM workspaces")
    void clearAllWorkspaces();
}

File: app/src/main/java/com/droidcode/db/WorkspaceDao.kt
package com.droidcode.db

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

@Dao
interface WorkspaceDao {

    @Query("SELECT * FROM workspaces ORDER BY lastOpenedTimestamp DESC")
    fun getAllWorkspacesFlow(): Flow<List<WorkspaceEntity>>

    @Query("SELECT * FROM workspaces ORDER BY lastOpenedTimestamp DESC")
    fun getAllWorkspacesSync(): List<WorkspaceEntity>

    @Query("SELECT * FROM workspaces WHERE path = :path LIMIT 1")
    fun getWorkspaceByPath(path: String): WorkspaceEntity?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertWorkspace(workspace: WorkspaceEntity)

    @Query("DELETE FROM workspaces WHERE path = :path")
    fun deleteWorkspaceByPath(path: String)

    @Query("DELETE FROM workspaces")
    fun clearAllWorkspaces()
}
